# Capstone Project

## 1. Description

*Determine the Problem*

For this assignment, we are required to define our own questions using the knowledge of API and location data to analyze the realistic problems. The ploblem I choose is how to choose a good restaurant and a place for fun for a student in University of Toronto to spend weekend time. It is because I am a university student now that this problem is highly correlated with my daily life, so solving this problems may be realistic to the sutdents like me. Moreover, as we clustered Toronto in previous weeks, I took university of Toronto as an example.

*Describe the Data*

The data required for this assignment can be acquired by Foursquare API--the geographical location and the ratings of restaurants. It is necessary to cluster Totonto and find the best place for students to go in Toronto. Wikipedia can provide the postal codes, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M and the geographical coordinates--latitutudes and longitudes of each neighborhood--can be acquired from the link,  http://cocl.us/Geospatial_data. The data should be cleaned and modified and then construct a dataframe to analyze the problem.

## 2. Evaluation of Restaurants

**2.1 Import necessary libraries**

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation


!pip install geopy
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize


! pip install folium==0.5.0
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


**2.2 Load and explore the data**

The centre is University of Toronto. Grocode is used to finding the geographical coordinates of University of Toronto.

In [2]:
address = "University of Toronto,Toronto, ON"

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of University of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of University of Toronto are 43.663461999999996, -79.39775965337452.


In [3]:
search_query = 'Italian'
radius = 50000000
print(search_query + ' .... OK!')

Italian .... OK!


In [ ]:
CLIENT_ID = # your Foursquare ID
CLIENT_SECRET = # your Foursquare Secret
ACCESS_TOKEN = # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Define URL of Foursquare and send the request.

In [5]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude,ACCESS_TOKEN, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=BTDHB1J1DB2RSBJ2PNVPGICOQOQQODHI2BMM2M1E5CJTEHMM&client_secret=VLAMETKFKYTMM3UNVYVCCDVN03RHLLS5LUBQZNBWRZEGT0CH&ll=43.663461999999996,-79.39775965337452&oauth_token=3UW1ZXW5TFKF23UMFOC0TEWV5PDEOY4HHESQLJ24VMC2CILB&v=20180604&query=Italian&radius=50000000&limit=30'

In [6]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fecc8b7d5239e00e2185601'},
 'notifications': [{'type': 'notificationTray', 'item': {'unreadCount': 0}}],
 'response': {'venues': [{'id': '4e31afdd091a973ec9c5a2b5',
    'name': 'Punto Gelato, Simply Italian',
    'location': {'address': '146 Cumberland St',
     'crossStreet': 'btwn Avenue Rd & Bay St',
     'lat': 43.66995452843031,
     'lng': -79.39260262435629,
     'labeledLatLngs': [{'label': 'display',
       'lat': 43.66995452843031,
       'lng': -79.39260262435629}],
     'distance': 833,
     'postalCode': 'M5R 1A8',
     'cc': 'CA',
     'neighborhood': 'Yorkville',
     'city': 'Toronto',
     'state': 'ON',
     'country': 'Canada',
     'formattedAddress': ['146 Cumberland St (btwn Avenue Rd & Bay St)',
      'Toronto ON M5R 1A8',
      'Canada']},
    'categories': [{'id': '4bf58dd8d48988d1c9941735',
      'name': 'Ice Cream Shop',
      'pluralName': 'Ice Cream Shops',
      'shortName': 'Ice Cream',
      'icon': {'prefix': 'https

In [7]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


id                                 name  \
0  4e31afdd091a973ec9c5a2b5         Punto Gelato, Simply Italian   
1  52f6816f11d24a43115dc834     Scaddabush Italian Kitchen & Bar   
2  5b897e92db1d81002c91df8c              Fabbrica Rustic Italian   
3  4db9bb6db5928d7fda728fc4       The Italian Cultural Institute   
4  4ad4c05ff964a52014f720e3  Papa Ceo Fine Italian Foods & Pizza   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d1c9941735', 'name': 'I...  v-1609353399    False   
1  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...  v-1609353399    False   
2  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...  v-1609353399    False   
3  [{'id': '4bf58dd8d48988d1e2931735', 'name': 'A...  v-1609353399    False   
4  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  v-1609353399    False   

            location.address     location.crossStreet  location.lat  \
0          146 Cumberland St  btwn Avenue Rd & Bay St     43.669955   
1  382 Yonge Street, Unit #7                  Gerrard     43.658920   
2         66 Wellington St W                      NaN     43.647161   
3               496 Huron St                      NaN     43.668432   
4         654 Spadina Avenue                      NaN     43.663300   

   location.lng                            location.labeledLatLngs  \
0    -79.392603  [{'label': 'display', 'lat': 43.66995452843031...   
1    -79.382891  [{'label': 'display', 'lat': 43.65892029202872...   
2    -79.381691  [{'label': 'display', 'lat': 43.647161, 'lng':...   
3    -79.402169  [{'label': 'display', 'lat': 43.66843245214709...   
4    -79.402317  [{'label': 'display', 'lat': 43.66329982959619...   

   location.distance location.postalCode location.cc location.neighborhood  \
0                833             M5R 1A8          CA             Yorkville   
1               1299             M5B 1S8          CA                   NaN   
2               2228             M5K 1E7          CA                   NaN   
3                657                 NaN          CA                   NaN   
4                367             M5S 2H7          CA                   NaN   

  location.city location.state location.country  \
0       Toronto             ON           Canada   
1       Toronto             ON           Canada   
2       Toronto             ON           Canada   
3       Toronto             ON           Canada   
4       Toronto             ON           Canada   

                           location.formattedAddress  
0  [146 Cumberland St (btwn Avenue Rd & Bay St), ...  
1  [382 Yonge Street, Unit #7 (Gerrard), Toronto ...  
2   [66 Wellington St W, Toronto ON M5K 1E7, Canada]  
3                 [496 Huron St, Toronto ON, Canada]  
4   [654 Spadina Avenue, Toronto ON M5S 2H7, Canada]

**Modify the Data**

In [8]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name            categories  \
0                Punto Gelato, Simply Italian        Ice Cream Shop   
1            Scaddabush Italian Kitchen & Bar    Italian Restaurant   
2                     Fabbrica Rustic Italian    Italian Restaurant   
3              The Italian Cultural Institute           Art Gallery   
4         Papa Ceo Fine Italian Foods & Pizza           Pizza Place   
5                   Italian Consulate Toronto   Embassy / Consulate   
6                     Elm Street Italian Deli    Italian Restaurant   
7                 Italian Chamber Of Commerce       Conference Room   
8                         john's italian cafe    Italian Restaurant   
9                   Rita's Italian Ice Canada        Ice Cream Shop   
10                  italian vegatarian  magic    Italian Restaurant   
11                          The Fresh Italian    Italian Restaurant   
12                       Italian Walk Of Fame         Historic Site   
13                   The Fresh Italian Eatery    Italian Restaurant   
14                      Bacaro Italian Eatery    Italian Restaurant   
15      Commisso Bros. & Racco Italian Bakery    Italian Restaurant   
16                   Mustachio Italian Eatery    Italian Restaurant   
17                      Gelato Simply Italian        Ice Cream Shop   
18    Italian Canadian Savings & Credit Union                  Bank   
19                       The Italian Sandwich          Trailer Park   
20                         LA's Italian + Bar    Italian Restaurant   
21     Classic italian style pizza food truck            Food Truck   
22               Guelph Italian Canadian Club          Voting Booth   
23                            Jamie's Italian    Italian Restaurant   
24     Giovanni's Italian Kitchen & Pizza Bar    Italian Restaurant   
25                 Gourment italian Sandwichs    Italian Restaurant   
26           Scaddabush Italian Kitchen & Bar    Italian Restaurant   
27                   Marinella Simply Italian    Italian Restaurant   
28                               Italian Cafe  Fast Food Restaurant   
29  Italian canadian savings and credit union              Building   

                            address              crossStreet        lat  \
0                 146 Cumberland St  btwn Avenue Rd & Bay St  43.669955   
1         382 Yonge Street, Unit #7                  Gerrard  43.658920   
2                66 Wellington St W                      NaN  43.647161   
3                      496 Huron St                      NaN  43.668432   
4                654 Spadina Avenue                      NaN  43.663300   
5                   136 Beverley St            Dundas Street  43.654027   
6                     15 Elm Street                      NaN  43.657690   
7                               NaN                      NaN  43.656158   
8                 27 Baldwin Street                      NaN  43.656127   
9                    347 College St                      NaN  43.657070   
10                 7 Yorkville Ave.                      NaN  43.673801   
11                              NaN                      NaN  43.654991   
12                              NaN                      NaN  43.655182   
13      109 McCaul Street, Unit #42       Dundas Street West  43.653889   
14                  2 Bloor St East            Yonge + Bloor  43.670824   
15                 8 Kincort Street  Castlefield and Kincort  43.697621   
16                       595 Bay St                Dundas St  43.656160   
17                    2076 Yonge St                      NaN  43.702080   
18          1168 Dundas Street West                      NaN  43.649369   
19                 242 Rushton Road                      NaN  43.680377   
20                              NaN                      NaN  43.650540   
21                       CNE Midway                      NaN  43.652144   
22                              NaN                      NaN  43.628227   
23                 3401 Dufferin St    

In [9]:
dataframe_filtered.name

0                  Punto Gelato, Simply Italian
1              Scaddabush Italian Kitchen & Bar
2                       Fabbrica Rustic Italian
3                The Italian Cultural Institute
4           Papa Ceo Fine Italian Foods & Pizza
5                     Italian Consulate Toronto
6                       Elm Street Italian Deli
7                   Italian Chamber Of Commerce
8                           john's italian cafe
9                     Rita's Italian Ice Canada
10                    italian vegatarian  magic
11                            The Fresh Italian
12                         Italian Walk Of Fame
13                     The Fresh Italian Eatery
14                        Bacaro Italian Eatery
15        Commisso Bros. & Racco Italian Bakery
16                     Mustachio Italian Eatery
17                        Gelato Simply Italian
18      Italian Canadian Savings & Credit Union
19                         The Italian Sandwich
20                           LA's Italia

**Visualize the nearby restaurants**

In [10]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the Conrad Hotel
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Conrad Hotel',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Italian restaurants as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

**2.3 Explore the restaurants further**

Let's see the ratings of the restaurants.

In [11]:
venue_id = '5b897e92db1d81002c91df8c' # ID of Fabbrica Rustic Italian
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


In [12]:
venue_id = '4ad4c05ff964a52014f720e3' # ID of Papa Ceo Fine Italian Foods & Pizza
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

5.2


The two restaurants are not so appealing.

In [13]:
venue_id = '52f6816f11d24a43115dc834' # ID of Scaddabush Italian Kitchen & Bar
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION)
url
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'price', 'hasMenu', 'likes', 'like', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'menu', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'description', 'page', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'hours', 'popular', 'seasonalHours', 'defaultHours', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto', 'colors'])
7.5


The rating of Scaddabush Italian Kitchen & Bar are quite good. Let's explore the tips.

In [14]:
result['response']['venue']['tips']['count']#get the number of tips.

70

In [15]:
## Scaddabush Italian Kitchen & Bar Tips
limit = 15 # set limit to be greater than or equal to the total number of tips
url = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&oauth_token={}&v={}&limit={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION, limit)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fecc8b9a58dfc2a7c612d1f'},
 'notifications': [{'type': 'notificationTray', 'item': {'unreadCount': 0}}],
 'response': {'tips': {'count': 67,
   'items': [{'id': '534cc79611d22c65ce90a382',
     'createdAt': 1397540758,
     'text': 'This is great place to come with friends or family. The servers are super sweet and the menu is delish! i ordered Penne Arrabiatta, but go for their pizzas. The small size for main course is huge!',
     'type': 'user',
     'canonicalUrl': 'https://foursquare.com/item/534cc79611d22c65ce90a382',
     'likes': {'count': 3,
      'groups': [{'type': 'others',
        'count': 3,
        'items': [{'id': '2730128',
          'firstName': 'Rahul',
          'lastName': 'Rishi',
          'gender': 'male',
          'countryCode': 'US',
          'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
           'suffix': '/2730128-VDBBQPPCKX2GFMUL.jpg'}},
         {'id': '6837140',
          'firstName': 'Gregory',
       

In [16]:
tips = results['response']['tips']['items']

tip = results['response']['tips']['items'][0]
tip.keys()

dict_keys(['id', 'createdAt', 'text', 'type', 'canonicalUrl', 'likes', 'like', 'logView', 'agreeCount', 'disagreeCount', 'todo', 'user'])

In [17]:
pd.set_option('display.max_colwidth', -1)
tips_df = json_normalize(tips) # json normalize tips
# columns to keep
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.id']
tips_filtered = tips_df.loc[:, filtered_columns]
# display tips
tips_filtered.reindex()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:1: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  if __name__ == '__main__':
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  from ipykernel import kernelapp as app


text  \
0  This is great place to come with friends or family. The servers are super sweet and the menu is delish! i ordered Penne Arrabiatta, but go for their pizzas. The small size for main course is huge!   

   agreeCount  disagreeCount                        id user.firstName  \
0  4           0              534cc79611d22c65ce90a382  Annie           

  user.lastName   user.id  
0  Kunwar        62346847

It is because we are using a personal developer account that we can access only 2 of the restaurant's tips, instead of all 15 tips.
Scaddabush Italian Kitchen & Bar is a good place to enjoy a meal with friends on weekend for the student in university of Toronto.

## 3.Evaluation of Places in Toronto

**3.1 Download the Dependencies**

In [18]:
import pandas as pd
import numpy as np 
import random
import requests

!pip install folium
import folium

!pip install lxml 

from bs4 import BeautifulSoup 

from sklearn.cluster import KMeans


from pandas.io.json import json_normalize


from geopy.geocoders import Nominatim


from IPython.display import Image 
from IPython.core.display import HTML
from IPython.display import display_html


import matplotlib.cm as cm
import matplotlib.colors as colors

Import the data from Wikipedia and transform them into pd dataframe.

**3.2 Load and Explore the Data**

In [19]:
source = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text
soup = BeautifulSoup(source, 'lxml')
table = soup.find("table")
table_rows = table.tbody.find_all("tr")

res = []
for tr in table_rows:
    td = tr.find_all("td")
    row = [tr.text for tr in td]
    
   
    if row != [] and row[1] != "Not assigned":
       
        if "Not assigned" in row[2]: 
            row[2] = row[1]
        res.append(row)

df=pd.DataFrame(res,columns=["PostalCode", "Borough", "Neighborhood"])
df.head()

PostalCode             Borough                 Neighborhood
0  M1A\n      Not assigned\n      Not assigned\n             
1  M2A\n      Not assigned\n      Not assigned\n             
2  M3A\n      North York\n        Parkwoods\n                
3  M4A\n      North York\n        Victoria Village\n         
4  M5A\n      Downtown Toronto\n  Regent Park, Harbourfront\n

Prepare the data.

In [20]:
df["Neighborhood"]=df["Neighborhood"].str.replace("\n","")
df["Borough"]=df["Borough"].str.replace("\n","")
df["PostalCode"]=df["PostalCode"].str.replace("\n","")
df.head(11)

PostalCode           Borough                                 Neighborhood
0   M1A        Not assigned      Not assigned                               
1   M2A        Not assigned      Not assigned                               
2   M3A        North York        Parkwoods                                  
3   M4A        North York        Victoria Village                           
4   M5A        Downtown Toronto  Regent Park, Harbourfront                  
5   M6A        North York        Lawrence Manor, Lawrence Heights           
6   M7A        Downtown Toronto  Queen's Park, Ontario Provincial Government
7   M8A        Not assigned      Not assigned                               
8   M9A        Etobicoke         Islington Avenue, Humber Valley Village    
9   M1B        Scarborough       Malvern, Rouge                             
10  M2B        Not assigned      Not assigned

In [21]:
df=df[df.Borough != 'Not assigned']
df=df.groupby(["PostalCode", "Borough"])["Neighborhood"].apply(", ".join).reset_index()
df['Neighborhood'] = np.where(df['Neighborhood'] == 'Not assigned',df['Borough'],df['Neighborhood'])
df.head(11)

PostalCode      Borough  \
0   M1B        Scarborough   
1   M1C        Scarborough   
2   M1E        Scarborough   
3   M1G        Scarborough   
4   M1H        Scarborough   
5   M1J        Scarborough   
6   M1K        Scarborough   
7   M1L        Scarborough   
8   M1M        Scarborough   
9   M1N        Scarborough   
10  M1P        Scarborough   

                                             Neighborhood  
0   Malvern, Rouge                                         
1   Rouge Hill, Port Union, Highland Creek                 
2   Guildwood, Morningside, West Hill                      
3   Woburn                                                 
4   Cedarbrae                                              
5   Scarborough Village                                    
6   Kennedy Park, Ionview, East Birchmount Park            
7   Golden Mile, Clairlea, Oakridge                        
8   Cliffside, Cliffcrest, Scarborough Village West        
9   Birch Cliff, Cliffside West                            
10  Dorset Park, Wexford Heights, Scarborough Town Centre

**load the geographical location**

In [22]:
Geo_df = pd.read_csv("https://cocl.us/Geospatial_data")
Geo_df.head()

Postal Code   Latitude  Longitude
0  M1B         43.806686 -79.194353
1  M1C         43.784535 -79.160497
2  M1E         43.763573 -79.188711
3  M1G         43.770992 -79.216917
4  M1H         43.773136 -79.239476

In [23]:
df_toronto = pd.merge(df, Geo_df, how='left', left_on = 'PostalCode', right_on = 'Postal Code')
df_toronto.drop("Postal Code", axis=1, inplace=True)
df_toronto.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0  M1B        Scarborough  Malvern, Rouge                          43.806686   
1  M1C        Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2  M1E        Scarborough  Guildwood, Morningside, West Hill       43.763573   
3  M1G        Scarborough  Woburn                                  43.770992   
4  M1H        Scarborough  Cedarbrae                               43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

**Map of Toronto**

In [24]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)
map_toronto

In [25]:
# add markers to map
for lat, lng, borough, neighborhood in zip(
        df_toronto['Latitude'], 
        df_toronto['Longitude'], 
        df_toronto['Borough'], 
        df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  

In [26]:
map_toronto

## Let's focus on the main part of Toronto.

In [27]:
df_toronto1 = df_toronto[df_toronto['Borough'].str.contains("Toronto")].reset_index(drop=True)
df_toronto1.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0  M4E        East Toronto     The Beaches                     43.676357   
1  M4K        East Toronto     The Danforth West, Riverdale    43.679557   
2  M4L        East Toronto     India Bazaar, The Beaches West  43.668999   
3  M4M        East Toronto     Studio District                 43.659526   
4  M4N        Central Toronto  Lawrence Park                   43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [28]:
map_toronto1 = folium.Map(location=[latitude, longitude], zoom_start=12)
for lat, lng, borough, neighborhood in zip(
        df_toronto1['Latitude'], 
        df_toronto1['Longitude'], 
        df_toronto1['Borough'], 
        df_toronto1['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto1)  
map_toronto1

In [29]:
neighborhood_name = df_toronto1.loc[0, 'Neighborhood']
print(f"The first neighborhood's name is '{neighborhood_name}'.")

The first neighborhood's name is 'The Beaches'.


In [30]:
neighborhood_latitude = df_toronto1.loc[0, 'Latitude'] 
neighborhood_longitude = df_toronto1.loc[0, 'Longitude'] 

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [31]:
# Get the top 100 venues that are in The Beaches within a radius of 500 meters
LIMIT = 100 
radius = 500 
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

In [32]:
results = requests.get(url).json()

In [33]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [34]:
# clean the json and structure it into a pandas dataframe
venues =  requests.get(url).json()["response"]['groups'][0]['items']

nearby_venues = json_normalize(venues) 


filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]


nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)


nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name         categories        lat        lng
0  Glen Manor Ravine                   Trail              43.676821 -79.293942
1  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
2  Grover Pub and Grub                 Pub                43.679181 -79.297215
3  Upper Beaches                       Neighborhood       43.680563 -79.292869

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):# create a function
    venues_list=[]
    
    
    for name, lat, lng in zip(names, latitudes, longitudes):
        
            
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [36]:
toronto1_venues = getNearbyVenues(names=df_toronto1['Neighborhood'],
                                   latitudes=df_toronto1['Latitude'],
                                   longitudes=df_toronto1['Longitude']
                                  )
toronto1_venues.head()

Neighborhood  Neighborhood Latitude  \
0  The Beaches                   43.676357               
1  The Beaches                   43.676357               
2  The Beaches                   43.676357               
3  The Beaches                   43.676357               
4  The Danforth West, Riverdale  43.679557               

   Neighborhood Longitude                               Venue  Venue Latitude  \
0 -79.293031               Glen Manor Ravine                   43.676821        
1 -79.293031               The Big Carrot Natural Food Market  43.678879        
2 -79.293031               Grover Pub and Grub                 43.679181        
3 -79.293031               Upper Beaches                       43.680563        
4 -79.352188               Pantheon                            43.677621        

   Venue Longitude     Venue Category  
0 -79.293942        Trail              
1 -79.297734        Health Food Store  
2 -79.297215        Pub                
3 -79.292869        Neighborhood       
4 -79.351434        Greek Restaurant

In [37]:
# check the number of venues of each neighborhood
toronto1_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                                                                        
Berczy Park                                                                                                 59                      
Brockton, Parkdale Village, Exhibition Place                                                                22                      
Business reply mail Processing Centre, South Central Letter Processing Plant Toronto                        15                      
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport  15                      
Central Bay Street                                                                                          62                      
Christie                                                                                                    16                      
Church and Wellesley                                                                                        79                      
Commerce Court, Victoria Hotel                                                                              100                     
Davisville                                                                                                  37                      
Davisville North                                                                                            8                       
Dufferin, Dovercourt Village                                                                                15                      
First Canadian Place, Underground city                                                                      100                     
Forest Hill North & West, Forest Hill Road Park                                                             4                       
Garden District, Ryerson                                                                                    100                     
Harbourfront East, Union Station, Toronto Islands                                                           100                     
High Park, The Junction South                                                                               24                      
India Bazaar, The Beaches West                                                                              21                      
Kensington Market, Chinatown, Grange Park                                                                   64                      
Lawrence Park                                                                                               3                       
Little Portugal, Trinity                                                                                    42                      
Moore Park, Summerhill East                                                                                 3                       
North Toronto West,  Lawrence Park                                                                          19                      
Parkdale, Roncesvalles                                                                                      14                      
Queen's Park, Ontario Provincial Government                                                                 30                      
Regent Park, Harbourfront                                                                                   48                      
Richmond, Adelaide, King                                                                                    96                      
Rosedale                                                                                                    4                       
Roselawn                                                                                                    2                       
Runnymede, Swansea                                                                                          34        

In [38]:
format(len(toronto1_venues['Venue Category'].unique()))

'235'

There are 235 uniques categories.

**3.3 Find the Proper Cluster**

In [39]:
toronto1_onehot = pd.get_dummies(toronto1_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto1_onehot['Neighborhood'] = toronto1_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto1_onehot.columns[-1]] + list(toronto1_onehot.columns[:-1])
toronto1_onehot = toronto1_onehot[fixed_columns]

toronto1_onehot.head()

Yoga Studio  Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0  0            0               0        0                   0              
1  0            0               0        0                   0              
2  0            0               0        0                   0              
3  0            0               0        0                   0              
4  0            0               0        0                   0              

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0  0               0                0                 0                     
1  0               0                0                 0                     
2  0               0                0                 0                     
3  0               0                0                 0                     
4  0               0                0                 0                     

   Antique Shop  ...  Theme Restaurant  Tibetan Restaurant  Toy / Game Store  \
0  0             ...  0                 0                   0                  
1  0             ...  0                 0                   0                  
2  0             ...  0                 0                   0                  
3  0             ...  0                 0                   0                  
4  0             ...  0                 0                   0                  

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  1      0              0                              0                  
1  0      0              0                              0                  
2  0      0              0                              0                  
3  0      0              0                              0                  
4  0      0              0                              0                  

   Vietnamese Restaurant  Wine Bar  Wine Shop  
0  0                      0         0          
1  0                      0         0          
2  0                      0         0          
3  0                      0         0          
4  0                      0         0          

[5 rows x 235 columns]

In [40]:
# group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
toronto1_grouped = toronto1_onehot.groupby('Neighborhood').mean().reset_index()
toronto1_grouped.head()

Neighborhood  \
0  Berczy Park                                                                                                  
1  Brockton, Parkdale Village, Exhibition Place                                                                 
2  Business reply mail Processing Centre, South Central Letter Processing Plant Toronto                         
3  CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport   
4  Central Bay Street                                                                                           

   Yoga Studio  Adult Boutique   Airport  Airport Food Court  Airport Gate  \
0  0.000000     0.0             0.000000  0.000000            0.000000       
1  0.000000     0.0             0.000000  0.000000            0.000000       
2  0.000000     0.0             0.000000  0.000000            0.000000       
3  0.000000     0.0             0.066667  0.066667            0.066667       
4  0.016129     0.0             0.000000  0.000000            0.000000       

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0  0.000000        0.000000         0.000000          0.0                   
1  0.000000        0.000000         0.000000          0.0                   
2  0.000000        0.000000         0.000000          0.0                   
3  0.133333        0.133333         0.133333          0.0                   
4  0.000000        0.000000         0.000000          0.0                   

   ...  Theme Restaurant  Tibetan Restaurant  Toy / Game Store  Trail  \
0  ...  0.0               0.0                 0.0               0.0     
1  ...  0.0               0.0                 0.0               0.0     
2  ...  0.0               0.0                 0.0               0.0     
3  ...  0.0               0.0                 0.0               0.0     
4  ...  0.0               0.0                 0.0               0.0     

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  0.0            0.016949                       0.0                
1  0.0            0.000000                       0.0                
2  0.0            0.000000                       0.0                
3  0.0            0.000000                       0.0                
4  0.0            0.016129                       0.0                

   Vietnamese Restaurant  Wine Bar  Wine Shop  
0  0.0                    0.000000  0.0        
1  0.0                    0.000000  0.0        
2  0.0                    0.000000  0.0        
3  0.0                    0.000000  0.0        
4  0.0                    0.016129  0.0        

[5 rows x 235 columns]

In [41]:
# 15 most common venues each
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 15

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto1_grouped['Neighborhood']

for ind in np.arange(toronto1_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto1_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0  Berczy Park                                                                                                  
1  Brockton, Parkdale Village, Exhibition Place                                                                 
2  Business reply mail Processing Centre, South Central Letter Processing Plant Toronto                         
3  CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport   
4  Central Bay Street                                                                                           

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Coffee Shop           Cocktail Bar          Farmers Market         
1  Café                  Coffee Shop           Breakfast Spot         
2  Light Rail Station    Park                  Garden Center          
3  Airport Terminal      Airport Lounge        Airport Service        
4  Coffee Shop           Italian Restaurant    Sandwich Place         

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0  Bakery                Beer Bar              Restaurant             
1  Pet Store             Stadium               Burrito Place          
2  Smoke Shop            Farmers Market        Fast Food Restaurant   
3  Sculpture Garden      Rental Car Location   Coffee Shop            
4  Café                  Burger Joint          Salad Place            

  7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
0  Seafood Restaurant    Cheese Shop           Eastern European Restaurant   
1  Restaurant            Climbing Gym          Performing Arts Venue         
2  Brewery               Burrito Place         Restaurant                    
3  Boat or Ferry         Harbor / Marina       Bar                           
4  Thai Restaurant       Bubble Tea Shop       Yoga Studio                   

  10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
0  Japanese Restaurant    Irish Pub              Shopping Mall           
1  Bakery                 Nightclub              Convenience Store       
2  Auto Workshop          Skate Park             Pizza Place             
3  Airport Gate           Airport Food Court     Airport                 
4  Indian Restaurant      Seafood Restaurant     Japanese Restaurant     

  13th Most Common Venue 14th Most Common Venue  15th Most Common Venue  
0  Indian Restaurant      Jazz Club              Hotel                   
1  Italian Restaurant     Intersection           Furniture / Home Store  
2  Gym / Fitness Center   Comic Shop             Discount Store          
3  Distribution Center    Ethiopian Restaurant   Escape Room             
4  Juice Bar              Korean Restaurant      Ramen Restaurant

**First, cluster 3 is not proper obviously because its most common venues are about airport. Cluster 0 is more like a residential area with different types of restaurants, pharmacies and markets. Second, cluster 1 and 2 are very similar with sports places such as gyms, stadiums and yoga studios, and different restaurants and bars. However, cluster 2 is more ideal because there are entertainments like  comic shop and skate park in cluster 2. What's more, venues in cluster 4 are very diversified too with vatious type of restaurants and snack stores. Thus, cluster 2 and cluster 4 are suitable to spend the time on weekend.**

Cluster the neighborhoods

In [42]:
kclusters = 5

toronto1_grouped_clustering = toronto1_grouped.drop('Neighborhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto1_grouped_clustering)

kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [43]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto1_merged = df_toronto1

toronto1_merged = toronto1_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto1_merged.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0  M4E        East Toronto     The Beaches                     43.676357   
1  M4K        East Toronto     The Danforth West, Riverdale    43.679557   
2  M4L        East Toronto     India Bazaar, The Beaches West  43.668999   
3  M4M        East Toronto     Studio District                 43.659526   
4  M4N        Central Toronto  Lawrence Park                   43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031  0               Pub                   Health Food Store      
1 -79.352188  0               Greek Restaurant      Coffee Shop            
2 -79.315572  0               Park                  Fast Food Restaurant   
3 -79.340923  0               Coffee Shop           Gastropub              
4 -79.388790  1               Park                  Bus Line               

  3rd Most Common Venue 4th Most Common Venue  ...   6th Most Common Venue  \
0  Trail                 Wine Shop             ...  Dessert Shop             
1  Italian Restaurant    Ice Cream Shop        ...  Furniture / Home Store   
2  Gym                   Sushi Restaurant      ...  Sandwich Place           
3  Bakery                Brewery               ...  American Restaurant      
4  Swim School           Event Space           ...  Escape Room              

     7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
0  Diner                    Discount Store               Distribution Center    
1  Fruit & Vegetable Store  Cosmetics Shop               Brewery                
2  Brewery                  Liquor Store                 Restaurant             
3  Music Store              Seafood Restaurant           Cheese Shop            
4  Electronics Store        Eastern European Restaurant  Donut Shop             

  10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
0  Donut Shop             Doner Restaurant       Deli / Bodega           
1  Bubble Tea Shop        Restaurant             Pub                     
2  Italian Restaurant     Burrito Place          Fish & Chips Shop       
3  Clothing Store         Pet Store              Park                    
4  Doner Restaurant       Dog Run                Distribution Center     

        13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  
0  Eastern European Restaurant  Electronics Store      Escape Room            
1  Café                         Caribbean Restaurant   Pizza Place            
2  Steakhouse                   Ice Cream Shop         Pub                    
3  Comfort Food Restaurant      Coworking Space        Convenience Store      
4  Discount Store               Diner                  Dessert Shop           

[5 rows x 21 columns]

In [44]:
#Visualization
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(
        toronto1_merged['Latitude'], 
        toronto1_merged['Longitude'], 
        toronto1_merged['Neighborhood'], 
        toronto1_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

**3.4 Check the Clusters**

In [45]:
toronto1_merged.loc[toronto1_merged['Cluster Labels'] == 0, toronto1_merged.columns[[1] + list(range(5, toronto1_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   East Toronto      0               Pub                    
1   East Toronto      0               Greek Restaurant       
2   East Toronto      0               Park                   
3   East Toronto      0               Coffee Shop            
5   Central Toronto   0               Gym                    
6   Central Toronto   0               Clothing Store         
7   Central Toronto   0               Pizza Place            
9   Central Toronto   0               Coffee Shop            
11  Downtown Toronto  0               Restaurant             
12  Downtown Toronto  0               Coffee Shop            
13  Downtown Toronto  0               Coffee Shop            
14  Downtown Toronto  0               Coffee Shop            
15  Downtown Toronto  0               Coffee Shop            
16  Downtown Toronto  0               Coffee Shop            
17  Downtown Toronto  0               Coffee Shop            
18  Downtown Toronto  0               Coffee Shop            
19  Downtown Toronto  0               Coffee Shop            
20  Downtown Toronto  0               Coffee Shop            
21  Downtown Toronto  0               Coffee Shop            
24  Central Toronto   0               Coffee Shop            
25  Downtown Toronto  0               Café                   
26  Downtown Toronto  0               Café                   
27  Downtown Toronto  0               Airport Terminal       
28  Downtown Toronto  0               Coffee Shop            
29  Downtown Toronto  0               Coffee Shop            
30  Downtown Toronto  0               Grocery Store          
31  West Toronto      0               Pharmacy               
32  West Toronto      0               Bar                    
33  West Toronto      0               Café                   
34  West Toronto      0               Mexican Restaurant     
35  West Toronto      0               Gift Shop              
36  West Toronto      0               Coffee Shop            
37  Downtown Toronto  0               Coffee Shop            
38  East Toronto      0               Light Rail Station     

            2nd Most Common Venue 3rd Most Common Venue  \
0   Health Food Store              Trail                  
1   Coffee Shop                    Italian Restaurant     
2   Fast Food Restaurant           Gym                    
3   Gastropub                      Bakery                 
5   Hotel                          Pizza Place            
6   Coffee Shop                    Yoga Studio            
7   Coffee Shop                    Dessert Shop           
9   Restaurant                     American Restaurant    
11  Bakery                         Café                   
12  Japanese Restaurant            Sushi Restaurant       
13  Café                           Park                   
14  Clothing Store                 Café                   
15  Café                           American Restaurant    
16  Cocktail Bar                   Farmers Market         
17  Italian Restaurant             Sandwich Place         
18  Café                           Restaurant             
19  Aquarium                       Café                   
20  Hotel                          Café                   
21  Restaurant                     Café                   
24  Sandwich Place                 Café                   
25  Bookstore                      Japanese Restaurant    
26  Vegetarian / Vegan Restaurant  Mexican Restaurant     
27  Airport Lounge                 Airport Service        
28  Seafood Restaurant             Restaurant             
29  Café                           Hotel                  
30  Café                           Park                   
31  Bakery                         Wine Shop              
32  Coffee Shop                    Asian Restaurant       
33  Coffee Shop                    Breakfast Spot         
34  Thai Restaurant                Café                   
35  Breakfast

In [46]:
toronto1_merged.loc[toronto1_merged['Cluster Labels'] == 1, toronto1_merged.columns[[1] + list(range(5, toronto1_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4  Central Toronto  1               Park                   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4  Bus Line              Swim School           Event Space            

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4  Ethiopian Restaurant  Escape Room           Electronics Store      

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
4  Eastern European Restaurant  Donut Shop            Doner Restaurant        

  11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
4  Dog Run                Distribution Center    Discount Store          

  14th Most Common Venue 15th Most Common Venue  
4  Diner                  Dessert Shop

In [47]:
toronto1_merged.loc[toronto1_merged['Cluster Labels'] == 2, toronto1_merged.columns[[1] + list(range(5, toronto1_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto  2               Lawyer                 

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8  Tennis Court          Trail                 Wine Shop              

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8  Doner Restaurant      Diner                 Discount Store         

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
8  Distribution Center   Dog Run               Donut Shop              

  11th Most Common Venue       12th Most Common Venue 13th Most Common Venue  \
8  Department Store       Eastern European Restaurant  Electronics Store       

  14th Most Common Venue 15th Most Common Venue  
8  Escape Room            Ethiopian Restaurant

In [48]:
toronto1_merged.loc[toronto1_merged['Cluster Labels'] == 3, toronto1_merged.columns[[1] + list(range(5, toronto1_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto  3               Park                   
23  Central Toronto   3               Park                   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10  Playground            Trail                 Deli / Bodega          
23  Trail                 Jewelry Store         Sushi Restaurant       

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10  Ethiopian Restaurant  Escape Room           Electronics Store      
23  Department Store      Ethiopian Restaurant  Escape Room            

          8th Most Common Venue        9th Most Common Venue  \
10  Eastern European Restaurant  Donut Shop                    
23  Electronics Store            Eastern European Restaurant   

   10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
10  Doner Restaurant       Dog Run                Distribution Center     
23  Donut Shop             Doner Restaurant       Dog Run                 

   13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  
10  Discount Store         Diner                  Dessert Shop           
23  Distribution Center    Discount Store         Diner

In [49]:
toronto1_merged.loc[toronto1_merged['Cluster Labels'] == 4, toronto1_merged.columns[[1] + list(range(5, toronto1_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto  4               Pool                   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22  Garden                Wine Shop             Deli / Bodega          

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22  Ethiopian Restaurant  Escape Room           Electronics Store      

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
22  Eastern European Restaurant  Donut Shop            Doner Restaurant        

   11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
22  Dog Run                Distribution Center    Discount Store          

   14th Most Common Venue 15th Most Common Venue  
22  Diner                  Dessert Shop

As we chose cluster 2 as the place to go, cluster 2 is in Central Toronto as the output shown above. Furthermore, cluster 4 is also a good option with different types of entertainments in Downtown Toronto.

## 4. Conclusion

**In conclusion, if a student in university of Toronto wants to spend the time on weekend, Scaddabush Italian Kitchen & Bar is recommended because of its close distance and good ratings. Then Central Toronto and Downtown Toronto are recommended to visit.**